File: scripts/build_site.py
from __future__ import annotations

import csv
import json
from datetime import datetime, timezone
from pathlib import Path

DATA_PATH = Path("data") / "records.csv"
DOCS_DATA_DIR = Path("docs") / "data"


def main() -> None:
    DOCS_DATA_DIR.mkdir(parents=True, exist_ok=True)

    with DATA_PATH.open("r", encoding="utf-8", newline="") as fh:
        reader = csv.DictReader(fh)
        records = [
            {
                "threadNumber": row.get("thread_number", ""),
                "metadentUpper": row.get("metadent_upper", ""),
                "metadentLower": row.get("metadent_lower", ""),
                "firstPostId": row.get("first_post_id", ""),
                "firstPostDateTime": row.get("first_post_datetime", ""),
            }
            for row in reader
            if row.get("thread_number")
        ]

    payload = {
        "updatedAt": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
        "totalRecords": len(records),
        "records": records,
    }

    (DOCS_DATA_DIR / "index.json").write_text(
        json.dumps(payload, ensure_ascii=False, indent=2) + "\n",
        encoding="utf-8",
    )
    (DOCS_DATA_DIR / "records.csv").write_text(DATA_PATH.read_text(encoding="utf-8"), encoding="utf-8")


if __name__ == "__main__":
    main()

File: scripts/build-site.mjs
import fs from "node:fs/promises";
import path from "node:path";

const DATA_PATH = path.resolve("data", "records.csv");
const DOCS_DATA_DIR = path.resolve("docs", "data");

async function main() {
  const raw = await fs.readFile(DATA_PATH, "utf8");
  const lines = raw.split(/\r?\n/).filter(Boolean);
  const records = lines.slice(1).map((line) => {
    const [threadNumber, metadentUpper, metadentLower, firstPostId, firstPostDateTime] = parseCsvLine(line);
    return {
      threadNumber,
      metadentUpper,
      metadentLower,
      firstPostId,
      firstPostDateTime
    };
  });

  await fs.mkdir(DOCS_DATA_DIR, { recursive: true });

  const indexPayload = {
    updatedAt: new Date().toISOString(),
    totalRecords: records.length,
    records: records.map((record) => ({
      threadNumber: record.threadNumber,
      metadentUpper: record.metadentUpper,
      metadentLower: record.metadentLower,
      firstPostId: record.firstPostId,
      firstPostDateTime: record.firstPostDateTime
    }))
  };

  await fs.writeFile(
    path.join(DOCS_DATA_DIR, "index.json"),
    JSON.stringify(indexPayload, null, 2) + "\n",
    "utf8"
  );
  await fs.copyFile(DATA_PATH, path.join(DOCS_DATA_DIR, "records.csv"));
}

function parseCsvLine(line) {
  const values = [];
  let current = "";
  let inQuotes = false;

  for (let index = 0; index < line.length; index += 1) {
    const character = line[index];

    if (character === '"') {
      if (inQuotes && line[index + 1] === '"') {
        current += '"';
        index += 1;
      } else {
        inQuotes = !inQuotes;
      }
      continue;
    }

    if (character === "," && !inQuotes) {
      values.push(current);
      current = "";
      continue;
    }

    current += character;
  }

  values.push(current);
  return values;
}

await main();
